Feature/kas 111 beslissingsfiche klaarmaken voor digitaal ondertekenen

StuyckRuben opened the pull request:
https://kanselarij.atlassian.net/browse/KAS-111

- [ ] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/416
- [ ] https://github.com/kanselarij-vlaanderen/digital-signing-service/pull/32
- [ ] https://github.com/kanselarij-vlaanderen/yggdrasil/pull/58

Commit:
Merge branch 'development' into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

Commit:
add route for beslissingen en notulen

Commit:
add option to create and delete signflows on decisions

Commit:
remove signflow editing not visible on decisions

Commit:
remove unused import

Commit:
create select-secretary-modal
edit create-sign-flow to only show signers

tdn committed:
Created filter on reports or minutes to create list for marked signflows

Commit:
added secretary select modal

Commit:
add functionality to create and manage signflows for decisions

sergiofenoll commented on app/components/documents/document-card.js:
```diff
@@ -172,9 +171,19 @@ export default class DocumentsDocumentCardComponent extends Component {
-    this.signMarkingActivity = yield this.args.piece.belongsTo('signMarkingActivity').reload();
-    const signSubcase = yield this.signMarkingActivity?.signSubcase;
-    this.signFlow = yield signSubcase?.signFlow;
+    const decisionActivity = yield this.args.piece.decisionActivity;
+    if (decisionActivity) {
+      const signFlow = yield this.store.queryOne('sign-flow', {
+        'filter[decision-activity][:id:]' : decisionActivity.get('id'),
+      });
+      this.signFlow = signFlow;
+      const signSubcase = yield this.signFlow.signSubcase;
+      this.signMarkingActivity = yield signSubcase.signMarkingActivity;
+    } else {
+      this.signMarkingActivity = yield this.args.piece.belongsTo('signMarkingActivity').reload();
+      const signSubcase = yield this.signMarkingActivity?.signSubcase;
+      this.signFlow = yield signSubcase?.signFlow;
+    }
```
I don't understand this if-else block. If the piece has a decision activity, why do we need to fetch the sign flow through it?

This will also give unexpected results when an agendaitem has multiple pieces (or a BF) that are marked to be signed (because this query will return a random sign flow, not necessarily the one related to the current piece.

Can't we just use the logic in the else branch for all pieces?

sergiofenoll committed:
feat: add cache-busting for sign-marking-activity to piece subtypes

The cache incorrectly doesn't clear the relationships for subtypes, it
seems, so this works around the cache so that we can correctly fetch the
data.

sergiofenoll committed:
chore: undo changes to fetching sign marking activivity of piece

Querying the sign marking activity as was done doesn't work because many
sign flows might be linked to the same decision activity. Go back to the
old behaviour which now works because we bypass the cache.

sergiofenoll committed:
chore: make the signature pill link to the correct signatures page

Regular pieces should still link to signatures.index, reports or minutes
should link to signatures.decisions

sergiofenoll committed:
chore: set the is-report-or-minutes flag for minutes

sergiofenoll committed:
feat: make subcase (and by extension case) optional when creating a sign flow

sergiofenoll committed:
fix: replace query params of signatures.decisions in controller

sergiofenoll committed:
chore: remove unused code from signature.decisions

For reports and minutes we won't have approvers or notifications so we
can outright remove that code.

sergiofenoll committed:
feat: deal with decision activity and meeting in signatures.ongoing

Minutes don't have a decision activity, instead we get the meeting date
and secretary directly from the meeting.

sergiofenoll committed:
feat: show meeting date in signature.ongoing-decisions

sergiofenoll committed:
chore: only show ongoing sign flows in signatures.decisions-ongoing

By filtering on the existence of a creator we ensure that the sign flow
has started.

sergiofenoll committed:
feat: filter out reports and minutes from signatures.ongoing

sergiofenoll committed:
feat: always show secretary panel in agendaitem

The existence of a subcase doesn't matter in this case, we always show
this panel.

sergiofenoll committed:
feat: split up Signatures::CreateSignFlow

The logic for signer selection of a regular piece vs a report or the
minutes is different enough that it makes more sense to split these off
entirely.

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

sergiofenoll committed:
feat: move calls to pdf generation into separate file

sergiofenoll committed:
Merge branch 'feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

sergiofenoll commented on app/components/agenda/agendaitem/agendaitem-decision.js:
```diff
@@ -8,6 +8,7 @@ import CONSTANTS from 'frontend-kaleidos/config/constants';
 import addLeadingZeros from 'frontend-kaleidos/utils/add-leading-zeros';
 import VRDocumentName from 'frontend-kaleidos/utils/vr-document-name';
 import { deleteFile } from 'frontend-kaleidos/utils/document-delete-helpers';
+import { generateDecisionReport } from 'frontend-kaleidos/utils/generate-pdf';
```
Noticed the new https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/services/decision-report-generation.js, will replace this with that ASAP, and probably add one for minutes generation/adapt the existing service to also generate the minutes.

kobemertens replied:
https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1887 I implemented the service for the minutes generation here

sergiofenoll committed:
Merge branch 'development' into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

sergiofenoll commented on app/utils/generate-pdf.js:
```diff
@@ -0,0 +1,56 @@
+async function generateDecisionReport(reportId) {
+  const response = await fetch(`/generate-decision-report/${reportId}`);
+  try {
+    const data = await response.json();
+    if (!response.ok) {
+      throw new Error(
+        `Backend response contained an error (status: ${response.status}): ${JSON.stringify(data)}`
+      );
+    }
+    return data
+  } catch (error) {
+    // Errors returned from services *should* still
+    // be valid JSON(:API), but we could encounter
+    // non-JSON if e.g. a service is down. If so,
+    // throw a nice error that only contains the
+    // response status.
+    if (error instanceof SyntaxError) {
+      throw new Error(
+        `Backend response contained an error (status: ${response.status})`
+      );
+    } else {
+      throw error;
+    }
+  }
+}
```
@ValenberghsSven I added these util functions before you added the `DecisionReportGeneration` service. I will start using the service in this PR as well, but I'm wondering if we shouldn't use the error handling here in the service? Right now the toaster (like the component code it got extracted from) only shows a toast with a generic error if something goes wrong, showing the error from the service might make debugging actual issues easier?

ValenberghsSven replied:
Made the service to avoid duplication, any error handling is welcome

sergiofenoll replied:
Added here https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1869/commits/9809bae3f2652520d1d1f07f7ba27973f39ee6ad

sergiofenoll committed:
feat: default newly created minutes to Intern Secretarie

sergiofenoll committed:
chore: allow everyone access to Notulen tab in agenda

Notulen should be propagated to everyone (but readability is based on
the access level), so we need to make the tab visible for everyone.

sergiofenoll committed:
style: hide bottom panel unless user can edit minutes

The panel is otherwise empty for all non-editors since we don't
propagate the piece parts.

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

ValenberghsSven committed:
KAS-111 add isReportOrMinutes to record and remove unused getter

sergiofenoll committed:
test: mark for signing can now always exist

sergiofenoll committed:
chore: add error handling from previous util to service

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-111-beslissingsfiche-klaarmaken-voor-digitaal-ondertekenen

ValenberghsSven commented on app/components/documents/document-preview/signatures-tab.hbs:
```diff
@@ -21,7 +21,7 @@
     </div>
-  {{else if this.agendaitem}}
+  {{else}}
     {{#if this.loadCanManageSignFlow.isRunning}}
```
this if check was removed, but further down in this component nothing changed?
Ex. a document of a meeting (agenda/documenten) now shows 
"De beslissing gekoppeld aan dit document is nog niet vrijgegeven" which doesn't make sense.

This If check needs to be less narrow but this might be too broad.

Decisions can't be marked in this view because of how we load the route.
` this.decisionActivity` is empty because we have not accounted for this being a treatment report instead of an agendaitem piece

ValenberghsSven commented on app/components/signature-pill.js:
```diff
@@ -57,7 +57,13 @@ export default class SignaturePillComponent extends Component {
+          route = "signatures.decisions";
+        } else {
+          route = "signatures.index";
+        }
```
isReportOrMinutes is a boolean property, should not need await.

Also, shorthanded:
`route = isReportOrMinutes ? "signatures.decisions" : "signatures.index"`

ValenberghsSven commented on app/components/signatures/create-sign-flow.hbs:
```diff
@@ -1,203 +1,13 @@
-    @onConfirm={{this.saveApprover}}
-    @onCancel={{toggle "showApproversModal" this}}
+{{#if @isReportOrMinute}}
+  <Signatures::CreateSignFlow::ReportOrMinute
```
`::ReportOrMinutes` ?

ValenberghsSven commented on app/config/permissions.js:
```diff
@@ -72,6 +72,7 @@ const groups = [
       'manage-users',
       'manage-alerts',
       'manage-minutes',
+      'manage-secretary-signatures',
```
Add a description in the list above

ValenberghsSven commented on app/services/decision-report-generation.js:
```diff
@@ -13,12 +13,36 @@ export default class DecisionReportGeneration extends Service {
-      this.toaster.error(this.intl.t('error-while-exporting-pdf'));
-      return;
+    try {
+      const response = await fetch(`/generate-decision-report/${report.id}`);
+      try {
+        const data = await response.json();
+        if (!response.ok) {
+          throw new Error(
+            `Backend response contained an error (status: ${response.status}): ${JSON.stringify(data)}`
```
guessing this could be the new `generate-pdf` util, but since we are moving action towards the backend I will check/change that in that PR

ValenberghsSven commented on app/services/signature-service.js:
```diff
@@ -94,43 +94,41 @@ export default class SignatureService extends Service {
+    // TODO: Shouldn't the short & long title be coming from the agendaitem. Also when would show or edit this data?
+    const signFlow = this.store.createRecord('sign-flow', {
+      openingDate: now,
+      shortTitle: _case?.shortTitle,
+      longTitle: _case?.title,
```
do we need a fallback for these 2 titles if there is no case? Also see TODO above, are these values ever used? Still wondering about that (also since current cases have no option for long titles)

ValenberghsSven commented on app/templates/agenda/minutes.hbs:
```diff
@@ -100,40 +100,36 @@
+                @closable={{false}}
+              />
+            {{else}}
+              <AuAlert
+                @title={{t "no-minutes-for-meeting"}}
+                @icon="circle-info"
+                @skin="info"
+                @size="small"
+                @closable={{false}}
+              />
```
so all legacy + meetings up until now wont have digital minutes, but they will have a relevant manual pdf in `/documents` (this meeting or the next, need to check). Out of scope for now.

ValenberghsSven commented on app/templates/agenda/minutes.hbs:
```diff
@@ -100,40 +100,36 @@
+              />
+            {{/if}}
+          </Auk::Panel::Body>
+          {{#if (and this.currentPiecePart.value (user-may "manage-minutes"))}}
```
`user-may` first? Second condition should not be checked if the first one is false

ValenberghsSven commented on app/templates/signatures/ongoing-decisions.hbs:
```diff
@@ -0,0 +1,94 @@
+  </m.sidebar>
+  <m.content @scroll={{true}}>
+    <KDataTable
+      data-test-route-search-publication-data-table
```
this and others. Don't copy test selectors, they should be unique.
Only if you edit tests and make a unique test selector (following some own custom rules) + update the selector files in the cypress folders is this ok.

ValenberghsSven commented on app/templates/agenda/minutes.hbs:
```diff
@@ -100,40 +100,36 @@
               <AuContent>
                 <SanitizeHtml
                   @raw={{true}}
                   @value={{this.currentPiecePart.value.value}}
```
out of scope. There are some minutes changes in this PR which is very confusing.
Also this `currentPiecePart.value.value` is not new, but it's very vague what we are doing.
Could we rename the value to something else like "htmlContent" because that is what we are doing.
tasks having a value and then that result having a value is hard to read.

ValenberghsSven replied:
follow up ticket made KAS-4354

ValenberghsSven commented on app/components/documents/document-card.js:
```diff
@@ -81,7 +81,6 @@ export default class DocumentsDocumentCardComponent extends Component {
       && this.signaturesEnabled
       && this.currentSession.may('manage-signatures')
-      && !!this.args.decisionActivity;
   }
```
this may have some side effects. Now you can mark any document for signing it seems (ex. meeting documents)

ValenberghsSven replied:
every doc is now markable, checks needs to be refined instead of removed

ValenberghsSven commented on app/components/signatures/create-sign-flow/piece.js:
```diff
@@ -103,15 +95,19 @@ export default class SignaturesCreateSignFlowComponent extends Component {
   @action
   saveApprover(approver) {
-    this.approvers.addObject(approver.toLowerCase());
+    this.approvers.addObject(approver);
```
this was to show the users the correct sanitized emails upon entering them.
When they type in capitals, we immediately change it to lowerCase so we can show that in the view.
example without
![image](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/assets/22245223/bb7b9711-af6d-45eb-9ccc-7b35496221dc)

ValenberghsSven commented on app/templates/signatures/ongoing-decisions.hbs:
```diff
@@ -0,0 +1,94 @@
+    <div class="au-c-sidebar">
+      <div class="au-c-sidebar__content">
+        <div
+          id="search-subroute-filters-area"
```
this is from the search route, does not belong here (also in ongoing route, must have been copied before)

ValenberghsSven commented on app/controllers/signatures/ongoing-decisions.js:
```diff
@@ -0,0 +1,103 @@
+  getDecisionActivityOrMeeting = async (signFlowOrPromise) => {
+    const signFlow = await signFlowOrPromise;
+    const decisionActivity = await signFlow.decisionActivity;
+    if (decisionActivity) {
+      return decisionActivity;
```
does this work? As far as I can tell we don't always get/set the correct `decisionActivity` when it is a report (like the signatures-tab in document preview)

ValenberghsSven commented on app/controllers/signatures/ongoing-decisions.js:
```diff
@@ -0,0 +1,103 @@
+    const record = await this.getDecisionActivityOrMeeting(signFlowOrPromise);
+
+    const modelName = record?.constructor?.modelName;
+    if (modelName === 'decision-activity') {
+      return record.startDate;
+    } else if (modelName === 'meeting') {
+      return record.plannedStart;
```
this is a semi copy. it is true that we initially set the startDate of the decision-activity with the plannedStart of the meeting but that one can be changed later and this startDate is never updated. (low chance, can happen)
Why not get meeting.plannedStart in both cases for accuracy?

ValenberghsSven commented on app/controllers/signatures/decisions.js:
```diff
@@ -0,0 +1,250 @@
+import { TrackedArray } from 'tracked-built-ins';
+import { PAGINATION_SIZES } from 'frontend-kaleidos/config/config';
+
+const MANDATORY_SORT_OPTION = 'decision-activity';
+const DEFAULT_SORT_OPTIONS = [
+  '-decision-activity.start-date',
+  MANDATORY_SORT_OPTION,
+  'sign-subcase.sign-marking-activity.piece.name',
+];
```
copied code, how do you sort on `decision-activity` and the `start-date` when it is not available for minutes.

ValenberghsSven commented on app/templates/signatures/decisions.hbs:
```diff
@@ -0,0 +1,210 @@
+    @numberOfItems={{@model.length}}
+    @totalNumberOfItems={{@model.meta.count}}
+    @onChangeSize={{fn (mut this.sizeSignaturesIndex)}}
+    @onChangePage={{fn (mut this.pageSignaturesIndex)}}
```
sizeSignaturesDecisions & pageSignaturesDecisions  instead

ValenberghsSven commented on app/components/signatures/select-secretary-modal.js:
```diff
@@ -0,0 +1,46 @@
+import { inject as service } from '@ember/service';
+import CONSTANTS from 'frontend-kaleidos/config/constants';
+
+export default class SignaturesSelectSecretaryModalComponent extends Component {
+  @service mandatees;
+
+  @tracked selectedSecretary;
+  @tracked secretaryOptions = [];
```
this model is not used anywhere. are you supposed to be able to select secretaries? Not implemented in the signatures.decisions route yet

ValenberghsSven replied:
implemented it in the mentioned route.

ValenberghsSven replied:
removed fully, edit of secretary not wanted

ValenberghsSven commented on app/components/signatures/create-sign-flow/report-or-minute.js:
```diff
@@ -0,0 +1,62 @@
+    const secretary = await head.secretary;
+
+    for (let decisionActivity of tail) {
+      const _secretary = await decisionActivity.secretary;
```
`decisionActivityOrMeeting` here to avoid future confusion

ValenberghsSven reviewed (commented):
review / will be fixing some of the comments made.

ValenberghsSven committed:
KAS-111 review comments

ValenberghsSven committed:
KAS-111 more review comments resolved
